src/Compilers/VisualBasic/Portable/Symbols/ConstraintsHelper.vb

```diff
@@ -1086,28 +1091,55 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
             Return New CompoundDiagnosticInfo(diagnostics)
         End Function
 
+        Public Enum ConstructorConstraintError
+            None
+            NoPublicParameterlessConstructor
+            MustInheritType
+            HasRequiredMembers
+        End Enum
+
         ''' <summary>
         ''' Return true if the class type has a public parameterless constructor.
         ''' </summary>
-        Public Function HasPublicParameterlessConstructor(type As NamedTypeSymbol) As Boolean
+        Public Function HasPublicParameterlessConstructor(type As NamedTypeSymbol, synthesizedIfMissing As Boolean) As ConstructorConstraintError
             type = type.OriginalDefinition
-            Debug.Assert(type.TypeKind = TypeKind.Class)
+            Debug.Assert(type.TypeKind = TypeKind.Class OrElse type.TypeKind = TypeKind.Structure)
 
             Dim sourceNamedType = TryCast(type, SourceNamedTypeSymbol)
 
-            If sourceNamedType IsNot Nothing AndAlso Not sourceNamedType.MembersHaveBeenCreated Then
+            If sourceNamedType IsNot Nothing AndAlso Not sourceNamedType.MembersHaveBeenCreated AndAlso sourceNamedType.TypeKind = TypeKind.Class Then
                 ' When we are dealing with group classes and synthetic entry points,
                 ' we can end up here while we are building the set of members for the type.
                 ' Using InstanceConstructors property will send us into an infinite loop.
-                Return sourceNamedType.InferFromSyntaxIfClassWillHavePublicParameterlessConstructor()
+                Debug.Assert(sourceNamedType.TypeKind = TypeKind.Class)
+                If sourceNamedType.InferFromSyntaxIfClassWillHavePublicParameterlessConstructor() Then
+                    Return ConstructorConstraintError.None
+                Else
+                    Return ConstructorConstraintError.NoPublicParameterlessConstructor
+                End If
             End If
 
+            Dim hasRequiredMembers = type.AllRequiredMembers.Count > 0
+
             For Each constructor In type.InstanceConstructors
                 If constructor.ParameterCount = 0 Then
-                    Return constructor.DeclaredAccessibility = Accessibility.Public
+                    If constructor.DeclaredAccessibility <> Accessibility.Public Then
+                        Return ConstructorConstraintError.NoPublicParameterlessConstructor
+                    ElseIf hasRequiredMembers AndAlso Not constructor.HasSetsRequiredMembers Then
+                        Return ConstructorConstraintError.HasRequiredMembers
+                    Else
+                        Return ConstructorConstraintError.None
+                    End If
                 End If
             Next
-            Return False
+
+            If Not synthesizedIfMissing Then
```

Comment:
>If Not synthesizedIfMissing Then [](http://example.com/codeflow?start=12&length=32)

It is not obvious why is this needed. Is VB not synthesizing default constructor for structs?